ipq806x:add support for COMFAST CF-E370AC

fall513 opened the pull request:
COMFAST CF-E370AC is a ceiling mount AP with PoE support, based on
Qualcomm/Atheros IPQ4029 + QCA8075.

Short specification:

    2x 10/100/1000 Mbps Ethernet, with PoE support
    256MB of RAM (DDR2)
    32 MB of FLASH (SPI)
    2T2R 2.4 GHz, 802.11b/g/n, wave 2
    2T2R 5 GHz, 802.11ac/n/a, wave 2
    built-in 4x 3 dBi antennas
    output power (max): 500 mW (27 dBm)
    1x RGB LED, 1x button
    built-in watchdog chipset

Flash instruction:

Original firmware is based on OpenWrt.
Use sysupgrade image directly in vendor GUI.

Signed-off-by: Ding Tengfei <dtf@comfast.cn>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://lede-project.org/submitting-patches

Please remove this message before posting the pull request.

robimarko commented:
Does upstream board-2.bin support BMI used in your device?

fall513 commented:
Hi @robimarko ，

> Does upstream board-2.bin support BMI used in your device?

No, but after change BMI to 22 and 23, WIFI work fine with my board.

robimarko commented:
Well,that is not really good.
You either need to include a custom board-2.bin that supports default BMI used by the device or get your BMI upstreamed.

fall513 commented:
Hi @robimarko ,

> Well,that is not really good.
You either need to include a custom board-2.bin that supports default BMI used by the device or get your BMI upstreamed.

OK, added board-2.bin for WiFi.

robimarko commented:
You have a conflict in the code since your PR is based on older version,OpenMesh A42 does not use ipq-wifi anymore.
Rebase your code on master branch

fall513 commented:
Updated.

chunkeey commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+			compatible = "qcom,tcsr";
+			reg = <0x194b000 0x100>;
+			qcom,usb-hsphy-mode-select = <TCSR_USB_HSPHY_HOST_MODE>;
+			status = "ok";
```
ah, this seems to be a 1:1 copy from 864-03-dts-ipq4019-dk01-add-tcsr-config-to-dtsi.patch, right?
the status = "ok"; can be omitted.

fall513 replied:
OK.

chunkeey commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+			qcom,wifi_noc_memtype_m0_m2 = <TCSR_WIFI_NOC_MEMTYPE_M0_M2>;
+		};
+
+		pinctrl@1000000 {
```
you can use the label reference &tlmm (see ipq4019.dtsi) and put this whole node after the soc node.
The same is true for the spi_0, blsp_dma, ... nodes.

chunkeey commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+		compatible = "gpio-leds";
+
+		wlan {
+			label = "comfast:blue:wlan";
```
from what I know, the label be "boardname:color:function". so what about "cf-e370ac:blue:[wlan|wan|lan]" for the label?

fall513 replied:
When wlan/wan/lan is working, the light blue/red/green is on.

chunkeey commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+
+		wlan {
+			label = "comfast:blue:wlan";
+			gpios = <&tlmm 0x0 GPIO_ACTIVE_HIGH>;
```
gpio-keys and the watchdog node both uses decimals for the gpio address. then again, this would be optional.

fall513 replied:
OK.

chunkeey commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -47,6 +47,7 @@ Don't install it for any other device!
   PREV_BOARD+=ipq-wifi-$(1)
 endef
 
+$(eval $(call generate-ipq-wifi-package,comfast_cf-e370ac,board-comfast_cf-e370ac.bin,COMFAST CF-E370AC))
```
Please also upstream your boardfile. The process is explained on the [ath10k wiki](https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles)

fall513 replied:
I had send email to ath10k@lists.infradead.org, but it reject my email with "host bombadil.infradead.org[65.50.211.133] said: 550 http://david.woodhou.se/email.html (in reply to end of DATA command)", I don't know reason.

robimarko replied:
Mailing list was not working a couple of times today.
Give it another go

chunkeey commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -11,6 +11,11 @@ board=$(board_name)
 
 case "$board" in
+comfast,cf-e370ac)
+	ucidef_set_led_wlan "wlan" "WLAN" "comfast:blue:wlan" "phy0tpt" "phy1tpt"
```
please change your label to follow the "boardname:color:function" rule.

fall513 replied:
OK.

fall513 commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
```
U-boot define it.

fall513 commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -17,6 +17,9 @@ avm,fritzbox-4040)
 	;;
+comfast,cf-e370ac)
+	ucidef_set_interfaces_lan_wan "eth0" "eth1"
+	;;
```
CF-E370AC is similar with openmesh a42, and it is not necessary configurate for switch.

fall513 commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+				compatible = "n25q128a11";
+				linux,modalias = "m25p80", "n25q128a11";
+				spi-max-frequency = <24000000>;
+				use-default-sizes;
```
OK.

fall513 commented on target/linux/ipq806x/base-files/lib/upgrade/comfast.sh:
```diff
@@ -0,0 +1,28 @@
+	echo "${INDEX:+$PREFIX$INDEX}"
+}
+
+platform_do_upgrade_comfast() {
```
That is not important, I will rename it.

fall513 commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+			pinctrl-0 = <&spi_0_pins>;
+			pinctrl-names = "default";
+			status = "ok";
+
```
OK

fall513 commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+					pins = "gpio54";
+					drive-strength = <2>;
+					bias-disable;
+					output-high;
```
OK

fall513 replied:
There must be something wrong, when I removed ouput-high, it can not initialize flash anymore.

fall513 commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+		wifi@a800000 {
+			status = "okay";
+		};
+		nand: qpic-nand@79b0000 {
```
OK

fall513 commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4029-cf-e370ac.dts:
```diff
@@ -0,0 +1,212 @@
+				#address-cells = <1>;
+				#size-cells = <1>;
+				reg = <0>;
+				compatible = "n25q128a11";
```
OK

cfpublish committed:
ipq806x:add support for COMFAST CF-E370AC

COMFAST CF-E370AC is a ceiling mount AP with PoE support, based on
Qualcomm/Atheros IPQ4029 + QCA8075.

Short specification:

    2x 10/100/1000 Mbps Ethernet, with PoE support
    256 MB of RAM (DDR2)
    32 MB of FLASH (SPI)
    2T2R 2.4 GHz, 802.11b/g/n, wave 2
    2T2R 5 GHz, 802.11ac/n/a, wave 2
    built-in 4x 3 dBi antennas
    output power (max): 500 mW (27 dBm)
    1x RGB LED, 1x button
    built-in watchdog chipset

Flash instruction:

Original firmware is based on OpenWrt.
Use sysupgrade image directly in vendor GUI.

Signed-off-by: Ding Tengfei <dtf@comfast.cn>

socketpair commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -88,6 +88,21 @@ define Device/avm_fritzbox-4040
+	DEVICE_TITLE := COMFAST CF-E370AC
+	IMAGES := sysupgrade.bin
+	IMAGE/sysupgrade.bin := append-kernel | pad-to $$$${KERNEL_SIZE} | append-rootfs | pad-rootfs | append-metadata
+	DEVICE_PACKAGES := ath10k-firmware-qca4019 ipq-wifi-comfast_cf-e370ac
```
is it typo? 4019 vs 4029

socketpair replied:
Also, is it ipq40xx device or ipq806x device ?

fall513 replied:
IPQ4029,

socketpair replied:
so, why you change  target/linux/**ipq806x**/image/Makefile ?
and why firmware called ath10k-firmware-**qca4019** ?

fall513 replied:
When I added this patch, ipq40xx is not created. And I will not update this path anymore, because it can used for developers with old OPENWRT/LEDE. Also I do not want to cost too much time on it.

socketpair replied:
Anyway, Did you test it on real hardware ? I want to continue your work, rebase to master and so on.

fall513 replied:
Yes, tested on real hardware,

socketpair commented:
Also, I have 21 such devices. And they have very strange bug about wifi performance speed from kilobits/s to 4 megabits/s. Different messages to dmesg. Did you check these changes in the real life ?
@fall513

fall513 commented:
@socketpair  My device's wifi performance speed is more than 200M/s. If you get a very low throughput result, please check the art partition.

klukonin commented:
@fall513 
Hello.
Did you test on this OpenWrt version or on your QSDK-based version with 3.14 kernel?
Also, could you please show the right variant of ART partition?


We have an issue when bit rates negotiate about 3-4 VHT MCS and up to 130 - 400+ Mbps, but throughput is about 200-300 kbps or 0.
Did you face with such behavior during you test?
And how can we fix ART table if it has wrong information on all CF-E370AC devices?

mkresin commented:
> When I added this patch, ipq40xx is not created. And I will not update this path anymore...

Than it shouldn't be keep as open PR.

@socketpair Feel free to open a new PR. Please take care of the discussions within the recently merged ipq40xx PRs like:

* https://github.com/openwrt/openwrt/commit/d0e962140403eefbe70002ee4349ade1d8c119be#diff-56084c5e657badcdff6ce29181444b61
* https://github.com/openwrt/openwrt/pull/810
* https://github.com/openwrt/openwrt/pull/713#issuecomment-369511149

to address these issue right from the beginning.

fall513 commented:
@klukonin 

> Did you test on this OpenWrt version or on your QSDK-based version with 3.14 kernel?

I tested on OpenWrt and QSDK, QSDK is a little better than OpenWrt.

> We have an issue when bit rates negotiate about 3-4 VHT MCS and up to 130 - 400+ Mbps, but throughput is about 200-300 kbps or 0.
Did you face with such behavior during you test?

No, throughput of my device is 200-500M/s with different mode.

> And how can we fix ART table if it has wrong information on all our CF-E370AC devices?

You should to re-calibrate ART table with IQxel.

socketpair commented:
@Stefan-Kolev Please tell me about you interest for these devices, why are you asking ?

Stefan-Kolev commented:
@socketpair I'm asking what happened with his PR. @fall513 got some suggestion about what to change to get the PR merged and there are no updates after that.
I'm interested because we are using them. We have a few hundred :)

fall513 commented:
@Stefan-Kolev , what's your question about CF-E370AC?

Stefan-Kolev commented:
@fall513 , My question is if you are going to make another PR to match your changes to the code to the latest OpenWRT and make official support for your device CF-E370AC?
As far as I understand this PR is now closed because some patches that you made are no longer compatible with latest version of OpenWRT.

fall513 commented:
@Stefan-Kolev , it's too messy to commit codes here. If you want to work with lastest version, I could make it with my forked codes.

klukonin commented:
@Stefan-Kolev 
We have our customized version of OpenWrt for this device.
The main goal of which is is nearby unauthorized Wi-Fi devices tracking.
It is not production ready, because we faced with some issues.
May be we can help each other and push a working version to the upstream.
For this time we support latest master with IPQ40xx target.
